### DISCUSSION TITLE
What are the exact rules to avoid namespace clashes when creating Oqtane modules?

### DISCUSSION THREAD

**Discussion by lanthonyneville** [oqtane/oqtane.framework#2570] [Q&A] [answered]
**Issue:** building my .Client project fails with multiple "_The type or namespace name 'Modules' does not exist in the namespace 'MyCompany.MyModule.Oqtane' (are you missing an assembly reference?)_" errors in _Imports.razor:

![image](https://user-images.githubusercontent.com/22200046/213933577-acf22b42-976f-40ea-b214-8f6a905eaa79.png)

**However** I have checked and checked again that none of the projects declare the namespace 'MyCompany.MyModule.Oqtane' (or anything under that). There is literally no mention of that namespace!

I thought I knew the rules (which are [mentioned in @iJungleboy 's article](https://azing.org/oqtane/r/IjQ147Ef) ) and I have created other modules and themes successfully, following these rules, but something about this one is causing an issue:

> Verify that your dll DOES contains the term Oqtane in the name
> Verify that your namespace DOES NOT contain the term Oqtane, unless it starts with that - so MyCompany.Oqtane will fail, Oqtane.MyCompany would work

What could be causing the clash??

FYI: using Oqtane 3.3.1 NuGets.

**Comment by iJungleboy** [comment]
Something in your code has an Oqtane namespace, after `MyCompany.MyModule.`
This could also just be a folder called `Oqtane` which in certain cases with Razor automatically provides the namespace for it.

I suggest you type `Oqtane.` and check what intellisense offers you after the dot. Then find that control and work your way forward from there.

**Comment by lanthonyneville** [comment, marked as answer]
Thanks @iJungleboy ... that pointed me to the issue. The intellisense showed me that Oqtane.Client._Imports was a valid class.

There is no namespace containing "Oqtane.Client" BUT the project name was MyCompany.MyModule.Oqtane.Client, and if I took "Oqtane" out of the project name, the issue disappeared, i.e. I made it MyCompany.MyModule.Client. I made sure to keep MyCompany.MyModule.Oqtane.Client as the assembly name however.

So, for the record, the rules to avoid namespace clashes (and have your assembly recognised) seem to be:

- Verify that your dll DOES contains the term Oqtane in the name
- Verify that your namespace DOES NOT contain the term Oqtane, unless it starts with that - so MyCompany.Oqtane will fail, Oqtane.MyCompany would work
- Verify that your client project DOES NOT contain any folders called Oqtane
- Verify that your client project DOES NOT contain the term Oqtane in the project name

**Reply by sbwalker** [reply]
The Module Creator could be enhanced to prevent developers for specifying "Oqtane" in the organization or module name - which would help avoid these types of issues.

**Reply by lanthonyneville** [reply]
> The Module Creator could be enhanced to prevent developers for specifying "Oqtane" in the organization or module name - which would help avoid these types of issues.

That would be great. I guess the issue has nothing specifically to do with Oqtane .... any Blazor project could have clashes with namespaces in referenced assemblies .... but as a newbie to Razor and  Blazor, those few hours trying to figure out what on earth was going on were painful :)
